src/Kestrel.Core/Internal/Http/Http1MessageBody.cs

```diff
@@ -130,16 +130,19 @@ private async Task PumpAsync()
             }
         }
 
-        public override Task StopAsync()
+        public override async Task StopAsync()
         {
             if (!_context.HasStartedConsumingRequestBody)
             {
-                return Task.CompletedTask;
+                return;
             }
 
             _canceled = true;
             _context.Input.CancelPendingRead();
-            return _pumpTask;
+            await _pumpTask;
```

Comment:
Let's use the two-method awaited pattern to avoid an extra async state machine in the normal case.